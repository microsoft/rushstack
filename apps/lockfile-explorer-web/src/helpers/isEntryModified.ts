// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type { ISpecChange } from '../parsing/compareSpec';
import type { LockfileEntry } from '../parsing/LfxGraph';

export const isEntryModified = (
  entry: LockfileEntry | undefined,
  specChanges: Map<string, ISpecChange>
): boolean => {
  if (!entry) return false;
  for (const dep of entry.dependencies) {
    if (specChanges.has(dep.name)) {
      return true;
    }
  }
  return false;
};
