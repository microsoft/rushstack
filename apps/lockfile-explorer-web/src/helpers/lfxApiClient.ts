// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type { IPackageJson } from '../types/IPackageJson';

const apiPath: string = `${window.appContext.serviceUrl}/api`;

export async function checkAliveAsync(): Promise<boolean> {
  try {
    await fetch(`${apiPath}/health`);
    return true;
  } catch (e) {
    return false;
  }
}

/**
 * Fetches a projects configuration files from the local file system
 *
 * @returns a json object representing a package.json or a text file to be rendered (in the case of readPnpmfile)
 */
export async function readPnpmfileAsync(): Promise<string> {
  try {
    const response = await fetch(`${apiPath}/pnpmfile`);
    return await response.text();
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error('Could not load cjs file: ', e);
    return 'Missing CJS';
  }
}

export async function readPackageJsonAsync(projectPath: string): Promise<IPackageJson | undefined> {
  try {
    const response = await fetch(`${apiPath}/package-json`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        projectPath
      })
    });
    return await response.json();
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error('Could not load package json file: ', e);
    return undefined;
  }
}

export async function readPackageSpecAsync(projectPath: string): Promise<IPackageJson | undefined> {
  try {
    const response = await fetch(`${apiPath}/package-spec`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        projectPath
      })
    });
    return await response.json();
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error('Could not load cjs file: ', e);
    return undefined;
  }
}
