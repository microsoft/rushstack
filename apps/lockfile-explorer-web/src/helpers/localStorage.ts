// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { type LockfileEntry, LockfileEntryFilter } from '../parsing/LfxGraph';

const BOOKMARK_KEY: string = 'LOCKFILE_EXPLORER_BOOKMARKS';

export const getBookmarksFromStorage = (): Set<string> => {
  const currBookmarks = JSON.parse(localStorage.getItem(BOOKMARK_KEY) || '{}');
  const bookmarkSet = new Set<string>();
  for (const key of Object.keys(currBookmarks)) {
    bookmarkSet.add(key);
  }
  return bookmarkSet;
};

export const saveBookmarkToLocalStorage = (entry: LockfileEntry): void => {
  const key = entry.rawEntryId;
  const currBookmarks = JSON.parse(localStorage.getItem(BOOKMARK_KEY) || '{}');
  currBookmarks[key] = true;
  localStorage.setItem(BOOKMARK_KEY, JSON.stringify(currBookmarks));
};

export const removeBookmarkFromLocalStorage = (entry: LockfileEntry): void => {
  const key = entry.rawEntryId;
  const currBookmarks = JSON.parse(localStorage.getItem(BOOKMARK_KEY) || '{}');
  delete currBookmarks[key];
  localStorage.setItem(BOOKMARK_KEY, JSON.stringify(currBookmarks));
};

const PROJECT_FILTER_KEY: string = 'LOCKFILE_EXPLORER_PROJECT_FILTER';
const PACKAGE_FILTER_KEY: string = 'LOCKFILE_EXPLORER_PACKAGE_FILTER';
export const saveFilterToLocalStorage = (filter: string, type: LockfileEntryFilter): void => {
  if (type === LockfileEntryFilter.Project) {
    localStorage.setItem(PROJECT_FILTER_KEY, filter);
  } else {
    localStorage.setItem(PACKAGE_FILTER_KEY, filter);
  }
};

export const getFilterFromLocalStorage = (type: LockfileEntryFilter): string => {
  if (type === LockfileEntryFilter.Project) {
    return localStorage.getItem(PROJECT_FILTER_KEY) || '';
  } else {
    return localStorage.getItem(PACKAGE_FILTER_KEY) || '';
  }
};
