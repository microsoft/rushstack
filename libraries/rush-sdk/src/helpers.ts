// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as path from 'node:path';

import { Import, FileSystem } from '@rushstack/node-core-library';
import type { EnvironmentVariableNames } from '@microsoft/rush-lib';

export const RUSH_LIB_NAME: '@microsoft/rush-lib' = '@microsoft/rush-lib';
export const RUSH_LIB_PATH_ENV_VAR_NAME: typeof EnvironmentVariableNames._RUSH_LIB_PATH = '_RUSH_LIB_PATH';

export type RushLibModuleType = Record<string, unknown>;

export interface ISdkContext {
  rushLibModule: RushLibModuleType | undefined;
}

export const sdkContext: ISdkContext = {
  rushLibModule: undefined
};

/**
 * Find the rush.json location and return the path, or undefined if a rush.json can't be found.
 *
 * @privateRemarks
 * Keep this in sync with `RushConfiguration.tryFindRushJsonLocation`.
 */
export function tryFindRushJsonLocation(startingFolder: string): string | undefined {
  let currentFolder: string = startingFolder;

  // Look upwards at parent folders until we find a folder containing rush.json
  for (let i: number = 0; i < 10; ++i) {
    const rushJsonFilename: string = path.join(currentFolder, 'rush.json');

    if (FileSystem.exists(rushJsonFilename)) {
      return rushJsonFilename;
    }

    const parentFolder: string = path.dirname(currentFolder);
    if (parentFolder === currentFolder) {
      break;
    }

    currentFolder = parentFolder;
  }

  return undefined;
}

export function _require<TResult>(moduleName: string): TResult {
  if (typeof __non_webpack_require__ === 'function') {
    // If this library has been bundled with Webpack, we need to call the real `require` function
    // that doesn't get turned into a `__webpack_require__` statement.
    // `__non_webpack_require__` is a Webpack macro that gets turned into a `require` statement
    // during bundling.
    return __non_webpack_require__(moduleName);
  } else {
    return require(moduleName);
  }
}

/**
 * Require `@microsoft/rush-lib` under the specified folder path.
 */
export function requireRushLibUnderFolderPath(folderPath: string): RushLibModuleType {
  const rushLibModulePath: string = Import.resolveModule({
    modulePath: RUSH_LIB_NAME,
    baseFolderPath: folderPath
  });

  return _require(rushLibModulePath);
}
