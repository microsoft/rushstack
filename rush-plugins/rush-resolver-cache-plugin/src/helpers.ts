// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { createHash } from 'node:crypto';
import * as path from 'node:path';

import type { ISerializedResolveContext } from '@rushstack/webpack-workspace-resolve-plugin';

import type { IDependencyEntry, IResolverContext } from './types';

const MAX_LENGTH_WITHOUT_HASH: number = 120 - 26 - 1;
const BASE32: string[] = 'abcdefghijklmnopqrstuvwxyz234567'.split('');

// https://github.com/swansontec/rfc4648.js/blob/ead9c9b4b68e5d4a529f32925da02c02984e772c/src/codec.ts#L82-L118
export function createBase32Hash(input: string): string {
  const data: Buffer = createHash('md5').update(input).digest();

  const mask: 0x1f = 0x1f;
  let out: string = '';

  let bits: number = 0; // Number of bits currently in the buffer
  let buffer: number = 0; // Bits waiting to be written out, MSB first
  for (let i: number = 0; i < data.length; ++i) {
    // eslint-disable-next-line no-bitwise
    buffer = (buffer << 8) | (0xff & data[i]);
    bits += 8;

    // Write out as much as we can:
    while (bits > 5) {
      bits -= 5;
      // eslint-disable-next-line no-bitwise
      out += BASE32[mask & (buffer >> bits)];
    }
  }

  // Partial character:
  if (bits) {
    // eslint-disable-next-line no-bitwise
    out += BASE32[mask & (buffer << (5 - bits))];
  }

  return out;
}

// https://github.com/pnpm/pnpm/blob/f394cfccda7bc519ceee8c33fc9b68a0f4235532/packages/dependency-path/src/index.ts#L167-L189
export function depPathToFilename(depPath: string): string {
  let filename: string = depPathToFilenameUnescaped(depPath).replace(/[\\/:*?"<>|]/g, '+');
  if (filename.includes('(')) {
    filename = filename.replace(/(\)\()|\(/g, '_').replace(/\)$/, '');
  }
  if (filename.length > 120 || (filename !== filename.toLowerCase() && !filename.startsWith('file+'))) {
    return `${filename.substring(0, MAX_LENGTH_WITHOUT_HASH)}_${createBase32Hash(filename)}`;
  }
  return filename;
}

/**
 * Computes the root folder for a dependency from a reference to it in another package
 * @param lockfileFolder - The folder that contains the lockfile
 * @param key - The key of the dependency
 * @param specifier - The specifier in the lockfile for the dependency
 * @param context - The owning package
 * @returns The identifier for the dependency
 */
export function resolveDependencyKey(
  lockfileFolder: string,
  key: string,
  specifier: string,
  context: IResolverContext
): string {
  if (specifier.startsWith('/')) {
    return getDescriptionFileRootFromKey(lockfileFolder, specifier);
  } else if (specifier.startsWith('link:')) {
    if (context.isProject) {
      return path.posix.join(context.descriptionFileRoot, specifier.slice(5));
    } else {
      return path.posix.join(lockfileFolder, specifier.slice(5));
    }
  } else if (specifier.startsWith('file:')) {
    return getDescriptionFileRootFromKey(lockfileFolder, specifier, key);
  } else {
    return getDescriptionFileRootFromKey(lockfileFolder, `/${key}@${specifier}`);
  }
}

/**
 * Computes the physical path to a dependency based on its entry
 * @param lockfileFolder - The folder that contains the lockfile during installation
 * @param key - The key of the dependency
 * @param name - The name of the dependency, if provided
 * @returns The physical path to the dependency
 */
export function getDescriptionFileRootFromKey(lockfileFolder: string, key: string, name?: string): string {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  if (!key.startsWith('file:')) {
    name = key.slice(1, key.indexOf('@', 2));
  }
  if (!name) {
    throw new Error(`Missing package name for ${key}`);
  }

  const originFolder: string = `${lockfileFolder}/node_modules/.pnpm/${depPathToFilename(key)}/node_modules`;
  const descriptionFileRoot: string = `${originFolder}/${name}`;
  return descriptionFileRoot;
}

export function resolveDependencies(
  lockfileFolder: string,
  collection: Record<string, IDependencyEntry>,
  context: IResolverContext
): void {
  for (const [key, value] of Object.entries(collection)) {
    const version: string = typeof value === 'string' ? value : value.version;
    const resolved: string = resolveDependencyKey(lockfileFolder, key, version, context);

    context.deps.set(key, resolved);
  }
}

/**
 *
 * @param depPath - The path to the dependency
 * @returns The folder name for the dependency
 */
export function depPathToFilenameUnescaped(depPath: string): string {
  if (depPath.indexOf('file:') !== 0) {
    if (depPath.startsWith('/')) {
      depPath = depPath.slice(1);
    }
    return depPath;
  }
  return depPath.replace(':', '+');
}

/**
 *
 * @param missingOptionalDependencies - The set of optional dependencies that were not installed
 * @param contexts - The map of context roots to their respective contexts
 * @param commonPathPrefix - The common root path to trim
 * @returns A function that serializes a context into a format that can be written to disk
 */
export function createContextSerializer(
  missingOptionalDependencies: Set<string>,
  contexts: Map<string, IResolverContext>,
  commonPathPrefix: string
): (entry: [string, IResolverContext]) => ISerializedResolveContext {
  return ([descriptionFileRoot, context]: [string, IResolverContext]): ISerializedResolveContext => {
    const { deps } = context;

    let hasAnyDeps: boolean = false;
    const serializedDeps: ISerializedResolveContext['deps'] = {};
    for (const [key, contextRoot] of deps) {
      if (missingOptionalDependencies.has(contextRoot)) {
        continue;
      }

      const resolutionContext: IResolverContext | undefined = contexts.get(contextRoot);
      if (!resolutionContext) {
        throw new Error(`Missing context for ${contextRoot}!`);
      }
      serializedDeps[key] = resolutionContext.ordinal;
      hasAnyDeps = true;
    }

    if (!context.name) {
      throw new Error(`Missing name for ${descriptionFileRoot}`);
    }

    const serializedContext: ISerializedResolveContext = {
      name: context.name,
      root: descriptionFileRoot.slice(commonPathPrefix.length),
      dirInfoFiles: context.nestedPackageDirs?.length ? context.nestedPackageDirs : undefined,
      deps: hasAnyDeps ? serializedDeps : undefined
    };

    return serializedContext;
  };
}
