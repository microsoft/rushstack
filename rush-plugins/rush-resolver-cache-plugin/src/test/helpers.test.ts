// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { createBase32Hash, depPathToFilename, getDescriptionFileRootFromKey } from '../helpers';

describe(createBase32Hash.name, () => {
  it('hashes', () => {
    for (const input of ['a', 'abracadabra', '(eslint@8.57.0)(typescript@5.4.5)']) {
      expect(createBase32Hash(input)).toMatchSnapshot(input);
    }
  });
});

describe(depPathToFilename.name, () => {
  it('formats', () => {
    for (const input of [
      '/autoprefixer@9.8.8',
      '/autoprefixer@10.4.18(postcss@8.4.36)',
      '/react-transition-group@4.4.5(react-dom@17.0.2)(react@17.0.2)',
      '/@some/package@1.2.3(@azure/msal-browser@2.28.1)(@azure/msal-common@6.4.0)(@fluentui/merge-styles@8.6.2)(@fluentui/react@8.117.5)(@fluentui/theme@2.6.45)(@fluentui/utilities@8.15.2)(chart.js@2.9.4)(lodash@4.17.21)(moment@2.29.4)(prop-types@15.8.1)(react-dnd-html5-backend@14.1.0)(react-dnd@14.0.5)(react-dom@17.0.1)(react-intersection-observer@8.34.0)(react@17.0.1)',
      '/@storybook/core@6.5.15(@storybook/builder-webpack5@6.5.15)(@storybook/manager-webpack5@6.5.15)(eslint@8.57.0)(react-dom@17.0.1)(react@17.0.1)(typescript@5.3.3)(webpack@5.88.1)',
      '/@typescript-eslint/utils@6.19.1(eslint@7.7.0)(typescript@5.4.2)',
      'file:../../../rigs/local-node-rig',
      'file:../../../libraries/ts-command-line(@types/node@18.17.15)'
    ]) {
      expect(depPathToFilename(input)).toMatchSnapshot(input);
    }
  });
});

describe(getDescriptionFileRootFromKey.name, () => {
  it('parses', () => {
    const lockfileRoot: string = '/$';
    for (const { key, name } of [
      { key: '/autoprefixer@9.8.8' },
      { key: '/autoprefixer@10.4.18(postcss@8.4.36)' },
      { key: '/react-transition-group@4.4.5(react-dom@17.0.2)(react@17.0.2)' },
      {
        key: '/@some/package@1.2.3(@azure/msal-browser@2.28.1)(@azure/msal-common@6.4.0)(@fluentui/merge-styles@8.6.2)(@fluentui/react@8.117.5)(@fluentui/theme@2.6.45)(@fluentui/utilities@8.15.2)(chart.js@2.9.4)(lodash@4.17.21)(moment@2.29.4)(prop-types@15.8.1)(react-dnd-html5-backend@14.1.0)(react-dnd@14.0.5)(react-dom@17.0.1)(react-intersection-observer@8.34.0)(react@17.0.1)'
      },
      {
        key: '/@storybook/core@6.5.15(@storybook/builder-webpack5@6.5.15)(@storybook/manager-webpack5@6.5.15)(eslint@8.57.0)(react-dom@17.0.1)(react@17.0.1)(typescript@5.3.3)(webpack@5.88.1)'
      },
      { key: '/@typescript-eslint/utils@6.19.1(eslint@7.7.0)(typescript@5.4.2)' },
      { key: 'file:../../../rigs/local-node-rig', name: 'local-node-rig' },
      {
        key: 'file:../../../libraries/ts-command-line(@types/node@18.17.15)',
        name: '@rushstack/ts-command-line'
      }
    ]) {
      expect(getDescriptionFileRootFromKey(lockfileRoot, key, name)).toMatchSnapshot(`"${key}",${name}`);
    }
  });
});
