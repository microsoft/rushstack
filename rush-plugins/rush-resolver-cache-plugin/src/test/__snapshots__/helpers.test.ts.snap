// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createBase32Hash hashes: (eslint@8.57.0)(typescript@5.4.5) 1`] = `"2nt7x5y4npubskzl4nixybkqqi"`;

exports[`createBase32Hash hashes: a 1`] = `"btaxlooa6g3kqmodthrgs5zgme"`;

exports[`createBase32Hash hashes: abracadabra 1`] = `"5rjiprc7bzyoyiwvf2f4x3vwia"`;

exports[`depPathToFilename formats: /@some/package@1.2.3(@azure/msal-browser@2.28.1)(@azure/msal-common@6.4.0)(@fluentui/merge-styles@8.6.2)(@fluentui/react@8.117.5)(@fluentui/theme@2.6.45)(@fluentui/utilities@8.15.2)(chart.js@2.9.4)(lodash@4.17.21)(moment@2.29.4)(prop-types@15.8.1)(react-dnd-html5-backend@14.1.0)(react-dnd@14.0.5)(react-dom@17.0.1)(react-intersection-observer@8.34.0)(react@17.0.1) 1`] = `"@some+package@1.2.3_@azure+msal-browser@2.28.1_@azure+msal-common@6.4.0_@fluentui+merge-style_yt7yh6tpppbzu7nx3lzx3f3ife"`;

exports[`depPathToFilename formats: /@storybook/core@6.5.15(@storybook/builder-webpack5@6.5.15)(@storybook/manager-webpack5@6.5.15)(eslint@8.57.0)(react-dom@17.0.1)(react@17.0.1)(typescript@5.3.3)(webpack@5.88.1) 1`] = `"@storybook+core@6.5.15_@storybook+builder-webpack5@6.5.15_@storybook+manager-webpack5@6.5.15__wnqw6tyxzeiksxiymflshxscdq"`;

exports[`depPathToFilename formats: /@typescript-eslint/utils@6.19.1(eslint@7.7.0)(typescript@5.4.2) 1`] = `"@typescript-eslint+utils@6.19.1_eslint@7.7.0_typescript@5.4.2"`;

exports[`depPathToFilename formats: /autoprefixer@9.8.8 1`] = `"autoprefixer@9.8.8"`;

exports[`depPathToFilename formats: /autoprefixer@10.4.18(postcss@8.4.36) 1`] = `"autoprefixer@10.4.18_postcss@8.4.36"`;

exports[`depPathToFilename formats: /react-transition-group@4.4.5(react-dom@17.0.2)(react@17.0.2) 1`] = `"react-transition-group@4.4.5_react-dom@17.0.2_react@17.0.2"`;

exports[`depPathToFilename formats: file:../../../libraries/ts-command-line(@types/node@18.17.15) 1`] = `"file+..+..+..+libraries+ts-command-line_@types+node@18.17.15"`;

exports[`depPathToFilename formats: file:../../../rigs/local-node-rig 1`] = `"file+..+..+..+rigs+local-node-rig"`;

exports[`getDescriptionFileRootFromKey parses: "/@some/package@1.2.3(@azure/msal-browser@2.28.1)(@azure/msal-common@6.4.0)(@fluentui/merge-styles@8.6.2)(@fluentui/react@8.117.5)(@fluentui/theme@2.6.45)(@fluentui/utilities@8.15.2)(chart.js@2.9.4)(lodash@4.17.21)(moment@2.29.4)(prop-types@15.8.1)(react-dnd-html5-backend@14.1.0)(react-dnd@14.0.5)(react-dom@17.0.1)(react-intersection-observer@8.34.0)(react@17.0.1)",undefined 1`] = `"/$/node_modules/.pnpm/@some+package@1.2.3_@azure+msal-browser@2.28.1_@azure+msal-common@6.4.0_@fluentui+merge-style_yt7yh6tpppbzu7nx3lzx3f3ife/node_modules/@some/package"`;

exports[`getDescriptionFileRootFromKey parses: "/@storybook/core@6.5.15(@storybook/builder-webpack5@6.5.15)(@storybook/manager-webpack5@6.5.15)(eslint@8.57.0)(react-dom@17.0.1)(react@17.0.1)(typescript@5.3.3)(webpack@5.88.1)",undefined 1`] = `"/$/node_modules/.pnpm/@storybook+core@6.5.15_@storybook+builder-webpack5@6.5.15_@storybook+manager-webpack5@6.5.15__wnqw6tyxzeiksxiymflshxscdq/node_modules/@storybook/core"`;

exports[`getDescriptionFileRootFromKey parses: "/@typescript-eslint/utils@6.19.1(eslint@7.7.0)(typescript@5.4.2)",undefined 1`] = `"/$/node_modules/.pnpm/@typescript-eslint+utils@6.19.1_eslint@7.7.0_typescript@5.4.2/node_modules/@typescript-eslint/utils"`;

exports[`getDescriptionFileRootFromKey parses: "/autoprefixer@9.8.8",undefined 1`] = `"/$/node_modules/.pnpm/autoprefixer@9.8.8/node_modules/autoprefixer"`;

exports[`getDescriptionFileRootFromKey parses: "/autoprefixer@10.4.18(postcss@8.4.36)",undefined 1`] = `"/$/node_modules/.pnpm/autoprefixer@10.4.18_postcss@8.4.36/node_modules/autoprefixer"`;

exports[`getDescriptionFileRootFromKey parses: "/react-transition-group@4.4.5(react-dom@17.0.2)(react@17.0.2)",undefined 1`] = `"/$/node_modules/.pnpm/react-transition-group@4.4.5_react-dom@17.0.2_react@17.0.2/node_modules/react-transition-group"`;

exports[`getDescriptionFileRootFromKey parses: "file:../../../libraries/ts-command-line(@types/node@18.17.15)",@rushstack/ts-command-line 1`] = `"/$/node_modules/.pnpm/file+..+..+..+libraries+ts-command-line_@types+node@18.17.15/node_modules/@rushstack/ts-command-line"`;

exports[`getDescriptionFileRootFromKey parses: "file:../../../rigs/local-node-rig",local-node-rig 1`] = `"/$/node_modules/.pnpm/file+..+..+..+rigs+local-node-rig/node_modules/local-node-rig"`;
